src/driver/linux_resource/efct_superbuf.c

```diff
@@ -77,17 +73,89 @@ static bool post_superbuf_to_app(struct efhw_nic_efct_rxq* q, struct efhw_efct_r
   return true;
 }
 
-static bool post_superbuf_to_apps(struct efhw_nic_efct_rxq* q)
+static bool drop_sbuf_ref(struct xlnx_efct_device *edev,
+                          struct xlnx_efct_client *client, int qid,
+                          struct efhw_nic_efct_rxq* q,
+                          int sbid)
+{
+  if( --q->superbuf_refcount[sbid] == 0 ) {
+    edev->ops->release_superbuf(client, qid, sbid);
+    --q->total_sbufs;
+    return true;
+  }
+  return false;
+}
+
+static uint32_t next_app_seq_min(struct efhw_efct_rxq *app,
+                         struct efhw_nic_efct_rxq* q,
+                         uint32_t min) {
+  if( SEQ_LT(app->next_sbuf_seq, q->sbufs.oldest_app_seq) ) {
+    app->shm->stats.skipped_bufs += q->sbufs.oldest_app_seq - app->next_sbuf_seq;
+    app->next_sbuf_seq = q->sbufs.oldest_app_seq;
+  }
+  return SEQ_MIN(app->next_sbuf_seq, min);
+}
+
+static void advance_oldest_app_seq(struct xlnx_efct_device *edev,
+                               struct xlnx_efct_client *client, int qid,
+                               struct efhw_nic_efct_rxq* q, uint32_t until) {
+  EFHW_ASSERT(SEQ_GE(until, q->sbufs.oldest_app_seq));
+  while(q->sbufs.oldest_app_seq != until) {
+    uint16_t sbid_to_drop = q->sbufs.q[q->sbufs.oldest_app_seq++ %
+                                       CI_ARRAY_SIZE(q->sbufs.q)].sbid;
+    drop_sbuf_ref(edev, client, qid, q, sbid_to_drop);
+  }
+}
+
+static void update_oldest_app_seq(struct xlnx_efct_device *edev,
+                              struct xlnx_efct_client *client, int qid,
+                              struct efhw_nic_efct_rxq* q)
 {
+  /* Update the oldest_app_seq pointer. When it advances, drop references
+   * to the sbufs it went past. */
+  uint32_t min = q->sbufs.added;
   struct efhw_efct_rxq *app;
+  for( app = q->live_apps; app; app = app->next ) {
+    min = next_app_seq_min(app, q, min);
```

Comment:
I'm concerned about drift in this loop, i.e. (truncating to a 16-bit world to make things easier to read) added=0x9000, app1=0x2000, app2=0xb000, app3=0x7000. That'd make min=0x7000 when going through this loop, which is clearly wrong. Need to think more about how that case is supposed to work and, more generally, what happens if an app is stopped for a very long time; it might be necessary to mark it as 'dead' some other way if it drops more than 2^31 behind. Given how long 2^31 is, I wouldn't be opposed to breaking such an app entirely and irrevocable, if that turns out to be easier to implement.

Reply:
Just to put we discussed into writing:
`q->sbufs.added - q->sbufs.oldest_app_seq <= CI_EFCT_MAX_SUPERBUFS << 2^31`
`next_app_seq_min` is called for each application each time an sbuf is posted to an app or when an app has finished with a superbuf and we succeed in posting another.
This function updates `app->next_sbuf_seq` to `q->sbufs.oldest_app_seq` whenever it is less than the oldest seq. 
This means the app's sequence numbers will never fall behind far enough to cause a drift.